aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -0,0 +1,482 @@
+      auto q_bias = qkv_bias_data[d + 0 * D];
+      auto k_bias = qkv_bias_data[d + 1 * D];
+      auto v_bias = qkv_bias_data[d + 2 * D];
+      auto q_data = qkv_data[b * _3D * T + t * _3D + d + 0 * D] + q_bias;
+      auto k_data = qkv_data[b * _3D * T + t * _3D + d + 1 * D] + k_bias;
+      auto v_data = qkv_data[b * _3D * T + t * _3D + d + 2 * D] + v_bias;
+      q_data = q_data * inv_sqrt_dim_per_head;
+      q_k_v_data
+          [0 * B * num_head * T * dim_per_head +
+           b * num_head * T * dim_per_head + nh * T * dim_per_head +
+           t * dim_per_head + dh] = q_data;
+      q_k_v_data
+          [1 * B * num_head * T * dim_per_head +
+           b * num_head * T * dim_per_head + nh * T * dim_per_head +
+           t * dim_per_head + dh] = k_data;
+      q_k_v_data
+          [2 * B * num_head * T * dim_per_head +
+           b * num_head * T * dim_per_head + nh * T * dim_per_head +
+           t * dim_per_head + dh] = v_data;
+    }
+  }
+}
+
+Tensor bmm_nt(const Tensor& a, const Tensor& b) {
+  auto a_ = a.view({a.size(0) * a.size(1), a.size(2), a.size(3)});
+  auto b_ = b.view({b.size(0) * b.size(1), b.size(2), b.size(3)});
+  auto bt_ = b_.transpose(2, 1);
+  auto c_ = at::bmm(a_, bt_);
+  return c_.view({a.size(0), a.size(1), a.size(2), b.size(2)});
+}
+
+void masked_softmax_dropout(
+    Tensor& attn_scores,
+    c10::optional<Tensor> attn_mask,
+    const Tensor& query) {
+  if (query.is_nested() && !attn_mask) {
+    // TODO: maybe we could do better than generating a mask every time?
+
+    attn_mask = NestedTensor_to_mask(query, 2);
+    // TODO: CPU path does not support transformer mask yet.
+    if (attn_scores.is_cpu()) {
+      attn_mask = attn_mask->view({-1, 1, 1, attn_scores.sizes()[3]});
+      // 1 means skip, 0 means keep.
+      // want:
+      // 0,0 -> 0
+      // 0,1 -> 1
+      // 1,1 -> 1
+      // so that's logical OR.
+      *attn_mask = *attn_mask | attn_mask->transpose(2, 3);
+      attn_mask = at::expand_inplace(attn_scores, *attn_mask)->contiguous();
+    }
+    attn_mask = attn_mask->to(query.device(), /*non-blocking=*/true);
+  }
+  if (attn_mask && attn_mask->dtype() != at::kBool) {
+    TORCH_WARN(
+        "Converting mask without torch.bool dtype to bool; this will "
+        "negatively affect performance. Prefer to use a boolean mask directly.");
+    attn_mask = attn_mask->to(at::kBool);
+  }
+  if (attn_scores.is_cpu() && attn_mask && attn_mask->dim() == 2) {
+    // TODO: CPU path does not support transformer mask yet.
+    const auto batch_size = attn_scores.sizes()[0];
+    const auto seq_len = attn_scores.sizes()[3];
+    TORCH_CHECK(attn_mask->sizes()[0] == batch_size);
+    TORCH_CHECK(attn_mask->sizes()[1] == seq_len);
+    attn_mask = attn_mask->view({batch_size, 1, 1, seq_len});
+    attn_mask = at::expand_inplace(attn_scores, *attn_mask)->contiguous();
+  }
+  if (attn_mask) {
+    attn_scores = _masked_softmax(attn_scores, *attn_mask);
```

Comment:
This doesn't look good, writing over mutable Tensor& ref instead of copying into it

Reply:
Oh maybe it's fine though, `attn_scores` gets passed in as an argument. Better not to also write over it though